src/bvals/bvals_var.cpp

```diff
@@ -142,74 +154,92 @@ void BoundaryVariable::CopyFluxCorrectionBufferSameProcess(NeighborBlock &nb, in
   // 1) which MeshBlock?
   MeshBlock &target_block = *pmy_mesh_->FindMeshBlock(nb.snb.gid);
   // 2) which element in vector of BoundaryVariable *?
-  BoundaryData<> *ptarget_bdata = &(target_block.pbval->bvars[bvar_index]->bd_var_flcor_);
+  BoundaryData<> *ptarget_bdata = &(target_block.pbval->bvars.at(label_)->bd_var_flcor_);
   target_block.deep_copy(ptarget_bdata->recv[nb.targetid], bd_var_flcor_.send[nb.bufid]);
   ptarget_bdata->flag[nb.targetid] = BoundaryStatus::arrived;
   return;
 }
 
 // Default / shared implementations of 4x BoundaryBuffer public functions
 
-//----------------------------------------------------------------------------------------
-//! \fn void BoundaryVariable::SendBoundaryBuffers()
-//  \brief Send boundary buffers of variables
-
-void BoundaryVariable::SendBoundaryBuffers() {
-  auto pmb = GetBlockPointer();
-  int mylevel = pmb->loc.level;
-  for (int n = 0; n < pmb->pbval->nneighbor; n++) {
-    NeighborBlock &nb = pmb->pbval->neighbor[n];
-    if (bd_var_.sflag[nb.bufid] == BoundaryStatus::completed) continue;
-    int ssize;
-    if (nb.snb.level == mylevel)
-      ssize = LoadBoundaryBufferSameLevel(bd_var_.send[nb.bufid], nb);
-    else if (nb.snb.level < mylevel)
-      ssize = LoadBoundaryBufferToCoarser(bd_var_.send[nb.bufid], nb);
-    else
-      ssize = LoadBoundaryBufferToFiner(bd_var_.send[nb.bufid], nb);
-    // fence to make sure buffers are loaded and ready to send
-    pmb->exec_space.fence();
-    if (nb.snb.rank == Globals::my_rank) {
-      // on the same process
-      CopyVariableBufferSameProcess(nb, ssize);
-    } else {
-#ifdef MPI_PARALLEL
-      PARTHENON_MPI_CHECK(MPI_Start(&(bd_var_.req_send[nb.bufid])));
-#endif
-    }
-
-    bd_var_.sflag[nb.bufid] = BoundaryStatus::completed;
-  }
-  return;
-}
-
 //----------------------------------------------------------------------------------------
 //! \fn bool BoundaryVariable::ReceiveBoundaryBuffers()
 //  \brief receive the boundary data
 
-bool BoundaryVariable::ReceiveBoundaryBuffers() {
+bool BoundaryVariable::ReceiveBoundaryBuffers(bool is_allocated) {
   bool bflag = true;
 
   auto pmb = GetBlockPointer();
   for (int n = 0; n < pmb->pbval->nneighbor; n++) {
     NeighborBlock &nb = pmb->pbval->neighbor[n];
+    if ((nb.snb.rank == Globals::my_rank) && !local_neighbor_allocated[n]) {
+      continue;
+    }
```

Comment:
No, we continue because the neighbor doesn't have this variable allocated and so it will never set our boundary status to arrived (on the same rank the sending block sets the boundary status to arrived).

However, I've changed this interaction now (so that we exactly match MPI results) and this looks the same now as before the PR.